src/dune_rules/dep_conf_eval.ml

```diff
@@ -40,12 +38,10 @@ let dep_on_alias_rec alias ~loc =
                   (Path.Source.to_string_maybe_quoted src_dir)
               ])
       }
-  | Some dir ->
+  | Some _ ->
     let name = Dune_engine.Alias.name alias in
-    let+ is_nonempty =
-      Action_builder.dep_on_alias_rec name (Context_name.of_string ctx_name) dir
-    in
-    if (not is_nonempty) && not (Alias.is_standard name) then
+    let+ alias_status = Alias_rec.dep_on_alias_rec name (Alias.dir alias) in
+    if alias_status = Missing && not (Alias.is_standard name) then
```

Comment:
One way to make this kind of code look a little more natural is to provide some helper functions from the `Alias_status` module.

Also, looking at this call site, I feel like maybe a better naming could be `Defined` and `Not_defined` (instead of `Exists` and `Missing`).

(But maybe renaming can happen in a separate PR, since we've already had "if_exists" before.)

Reply:
> One way to make this kind of code look a little more natural is to provide some helper functions from the Alias_status module.

I've become a little skeptical of that after working with dune. All these little helpers that don't contain any logic are painful to read because of all the jumping back and forth. They often all need to be all removed in the next iteration of refactoring as well so it's all additional churn.

I think a lot of the issues you've pointed in this PR wouldn't have occurred in the first place if we weren't so eager to generalize and introduce API's without thinking them through.

Reply:
I switched it to Defined/Not_defined and replaced the polymorphic comparison with a pattern match.

Reply:
Cool, thanks! I agree that over-generalisation is bad and we should get rid of it.

At the same time, my experience tells me that turning any non-trivial type into a module always pays off in the end. I think it has nothing to do with over-generalisation: it just gives helper functions a natural home (and most non-trivial types start needing helpers earlier or later, for better ergonomics).

Reply:
Right but a function like `Alias_status.is_defined` is definitely bordering the trivial.